### ISSUE TITLE
Binding Event Order Issue

### ISSUE THREAD

**Issue by grantcarthew** [aurelia/framework#371]
### Senario

I have a `view` that has a `custom element` in it with a `@binding` and I need the `view-model` to cache data to localStorage when the `custom element` changes.

The `custom element` contains input checkboxes that are bound `checked.bind='object.boolproperty'` and also have `change.delegate='checkboxClicked()'`.

I am using `element.dispatchEvent` on the `custom element` to inform the hosting view that the checkboxes have changed. The hosting view then saves the changes using a `cacheForm()` function on the `view-model`.

### The Problem

The `cacheForm()` function on the `view-model` in the parent `view` is caching the form data from the `custom element` to local store before the `@binding` value has updated the data.

### Question

How do I make sure the `view-model` bound property that is being passed to the `custom element` has been updated by changes on the `custom element` prior to caching the data?

Am I doing this all wrong?

### What have I tried?

*   debounce
*   debounce:1000
*   signal
*   valueChanged

### Code

Here's some code to add some clarity to my question.

A service class I am using to raise the `change` event on the `custom element`

```js
import {LogManager} from 'aurelia-framework'
var logger = LogManager.getLogger('service-event')

export class ServiceEvent {
  raiseChangeEvent (element) {
    logger.debug('raiseChangeEvent')
    let changeEvent
    if (window.CustomEvent) {
      changeEvent = new CustomEvent('change', {
        bubbles: true
      })
    } else {
      changeEvent = document.createEvent('CustomEvent')
      changeEvent.initCustomEvent('change', true, true)
    }
    element.dispatchEvent(changeEvent)
  }
}
```

The `custom element` view html containing the checked binding and change delegate.

```html
<ul repeat.for="bf of fields.BroadField" class="fields-list-checkbox">
  <li>
    <input id="bf-${bf.RefNo}" type="checkbox" model.bind="bf" checked.bind="bf.checked" change.delegate="checkboxClicked()"/>
    <label for="bf-${bf.RefNo}"><span title="Broad Field">${bf.Name}</span></label>
  </li>
</ul>
```

In the `custom element` `view model` within the `checkboxClicked()` function I am making these two calls:

```js
// this.selectedFields is the @bindable property.
// It gets updated with updatedFields after processing the selected option.
this.selectedFields = updatedFields 
// Notice I am raising the change event after the bound property has been updated.
this.serviceEvent.raiseChangeEvent(this.element)
```

The cacheForm function contained in the parent `view-model` (nothing to see here):

```js
  cacheForm () {
    logger.debug('cacheForm')
    window.localStorage.setItem('newActivity', JSON.stringify(this.activity))
  }
```

**Comment by grantcarthew** [comment]
Oh, I just found this on [reddit - Aurelia Observer Patterns](https://github.com/axwalker/aurelia-observer-patterns/wiki/Observer-patterns-in-Aurelia). I will try each one and see if I can get a solution. Thanks to @axwalker.

**Comment by grantcarthew** [comment]
OK, so I fixed my issue thanks to the document @axwalker wrote. Although I think it is a little hackish. I added a bindable value on the parent and changed the value on the child custom element when the element changed. I made the bindable value an integer and just increment it each time a checkbox on the child is checked and processed.

This worked over DOM events, EventAggregator and other methods I used.

There needs to be a better supported way of raising events on the custom element AFTER the bindings have been updated on the Parent. I don't know how this could be achieved.

I'll let someone else close this. I don't think it should be closed until a better solution is reached. Ultimately, a child element should be able to raise Changed events through the DOM and let the parent know it has Changed.

Pinging you @jdanyow fyi.

As always, thanks for all your hard work guys. Loving Aurelia.
